src/engine/strat_engine/devlinks.rs

```diff
@@ -2,170 +2,266 @@
         }
+    }
+
+    pub fn filesystem_renamed(pool_name: &str, fs_name: &str, _: &str) {
+        fn trigger_udev(pool_name: &str, fs_name: &str) -> StratisResult<()> {
+            let path: PathBuf = [DEV_PATH, pool_name, fs_name].iter().collect();
+
+            let dm_path = path.canonicalize()?;
+            let file_name = dm_path
+                .file_name()
+                .and_then(|s| s.to_str())
+                .ok_or_else(|| {
+                    StratisError::Engine(
+                        ErrorEnum::NotFound,
+                        format!(
+                            "Failed to generate a change uevent for {} to \
+                            make /dev/stratis symlinks consistent with internal pool \
+                            state",
+                            dm_path.display()
+                        ),
+                    )
+                })?;
+
+            let uevent_path: PathBuf = [UEVENT_PATH, file_name, "uevent"].iter().collect();
 
-        for leftover in existing_files {
-            filesystem_removed(pool_name, &leftover);
+            let mut uevent_file = OpenOptions::new().write(true).open(&uevent_path)?;
+            uevent_file.write_all(UEVENT_CHANGE_EVENT)?;
+
+            Ok(())
         }
 
-        Ok(())
-    }() {
-        warn!(
-            "setup_pool_devlinks failed for /stratis/{}, reason {:?}",
-            pool_name, err
-        );
-    };
+        if let Err(e) = trigger_udev(pool_name, fs_name) {
+            warn!(
+                "Synthetic udev event was not able to be triggered: {}. Rename of \
+                filesystem link {} failed",
+                e, fs_name,
+            );
+        }
+    }
 }
 
-/// Clean up directories and symlinks under /stratis based on current
-/// config. Clear out any directory or file that doesn't correspond to a pool.
-// Don't just remove everything in case there are processes
-// (e.g. user shells) with the current working directory within the tree.
-pub fn cleanup_devlinks<'a, I: Iterator<Item = (&'a Name, &'a PoolUuid, &'a StratPool)>>(pools: I) {
-    if let Err(err) = || -> StratisResult<()> {
-        let mut existing_dirs = fs::read_dir(DEV_PATH)?
-            .map(|dir_e| dir_e.map(|d| d.file_name().into_string().expect("Unix is utf-8")))
-            .collect::<Result<HashSet<_>, _>>()?;
-
-        for (pool_name, _, _) in pools {
-            existing_dirs.remove(&pool_name.to_owned());
-        }
+#[cfg(test)]
+mod test_methods {
+    use super::*;
+
+    use std::{collections::HashSet, io::ErrorKind, os::unix::fs::symlink, path::Path};
+
+    use crate::engine::{
+        engine::Pool,
+        strat_engine::pool::StratPool,
+        types::{Name, PoolUuid},
+    };
 
-        for leftover in existing_dirs {
-            pool_removed(&Name::new(leftover));
+    /// Set up the root Stratis directory, where dev links as well as temporary
+    /// MDV mounts will be created. This must occur before any pools are setup.
```

Comment:
I believe that the statement about temporary MDV mounts will become wholely false with this change, so it might as well be removed.